### ISSUE TITLE
Data couln't be read because it isn't in the correct format has cost me so much in wasted OPenAI Credit

### ISSUE THREAD

**Issue by bwt615** [BoltAI/BoltAI#112]
![CleanShot 2024-06-20 at 11 49 59@2x](https://github.com/BoltAI/BoltAI/assets/48396397/0914ef79-d355-4db2-944f-eace57f6ad29)

I get this like 85/90% of the time with tool use

**Comment by longseespace** [comment]
This usually happens when the AI model calls the function with a malformed/incorrect parameter.

It happens more frequently with Anthropic Claude or models other than GPT-4.

I've improved the error handling in the latest version but unfortunately it still largely depends on the AI model's Function Calling capability. 

Can you update to the latest version and let me know if it's improved. 

Thanks

**Comment by longseespace** [comment]
Interesting. Can you export the chat as a json so I could investigate further. 

Also which AI service/model/plugins are you using?

When this happens, there is no answer from LLM yet, but it's still in the progress of executing the tools. 

I can save the output of the tool calls but then it would still result in a failure as the parameters are incorrect. 

Which is not very helpful.

Most other apps choose to send the error back to the LLM and the whole process of tool use restarts.

This seems to be the best solution for now, but it would cost more tokens.

**Comment by bwt615** [comment]
Well I wonder, if the user has to resubmit their question anyway (or click regenerate), wouldn't that actually be more expensive than you just sending the error message back to the LLM? That's what I've been finding, especially in the cases where you have a convo cut off and it loses a key piece of context because of the new message.

GPT 4o, using bash

[chat-eda6a303-6280-4e54-84ba-ceb49ce80942.json](https://github.com/user-attachments/files/15917200/chat-eda6a303-6280-4e54-84ba-ceb49ce80942.json)

**Comment by longseespace** [comment]
Yeah currently BoltAI also attempts to send back the error to the LLM. But in some cases it might misses the decoding error (especially when trying to parse the back-and-forth function params)

I'm taking a deeper look right now.

**Comment by longseespace** [comment]
I found the issue. The function arguments from GPT-4o is not in valid JSON. I tried with a few different prompts but it still struggles to come up with a valid one.

And if I send the error to the LLM, it retries over and over, which can be quite expensive.

I think we've hit the limit of the model here and probably need another solution for this.

I'm working on the official Assistant API from OpenAI so we can use Code Interpreter for these kind of tasks.

**Comment by bwt615** [comment]
Could you also pass that to AI somehow and ask it to correct the JSON? xD


[Benjamin Thomas]

Benjamin Thomas

Mobile: +1 (781) 864-3648 | Fax: +1 (781) 787-2609

Email: ***@***.******@***.***> | Book: My Calendar<http://book.bthomas.io/> | Website: BThomas.io<http://bthomas.io/>

Twitter: @benjaminwthomas<https://twitter.com/benjaminwthomas> | LinkedIn: benjaminwesleythomas<https://www.linkedin.com/in/benjaminwesleythomas/>



From: Daniel Nguyen ***@***.***>
Date: Thursday, June 20, 2024 at 1:42 PM
To: BoltAI/BoltAI ***@***.***>
Cc: Benjamin Thomas ***@***.***>, Author ***@***.***>
Subject: Re: [BoltAI/BoltAI] Data couln't be read because it isn't in the correct format has cost me so much in wasted OPenAI Credit (Issue #112)


I found the issue. The function arguments from GPT-4o is not in valid JSON. I tried with a few different prompts but it still struggles to come up with a valid one.

And if I send the error to the LLM, it retries over and over, which can be quite expensive.

I think we've hit the limit of the model here and probably need another solution for this.

I'm working on the official Assistant API from OpenAI so we can use Code Interpreter for these kind of tasks.
